agent/src/util/device_plugin_service.rs

```diff
@@ -1637,7 +2233,848 @@ mod device_plugin_service_tests {
+        let kube_instance = kube_instance_builder.build();
+        let mock_instances = vec![(instance_name.to_string(), kube_instance)];
+        let mut mock = MockKubeInterface::new();
+        setup_find_instance_with_mock_instances(&mut mock, instance_namespace, mock_instances);
+
+        let device_usage_state = get_instance_device_usage_state(
+            node_name,
+            instance_name,
+            instance_namespace,
+            &capacity,
+            Arc::new(mock),
+        )
+        .await;
+        assert!(device_usage_state
+            .into_iter()
+            .all(|(_, v)| { v == DeviceUsageStatus::Free }));
+    }
+
+    // Tests correct device usage is returned when an Instance is not found
+    // Expected behavior: should return DeviceUsageStatus::Unknown for all usage slots
+    #[tokio::test]
+    async fn test_get_instance_device_usage_state_no_instance() {
+        let _ = env_logger::builder().is_test(true).try_init();
+        let node_name = "node-a";
+        let instance_name = "instance-1";
+        let instance_namespace = "test-namespace";
+        let capacity = 5i32;
+        let mut mock = MockKubeInterface::new();
+        setup_find_instance_with_not_found_err(&mut mock, instance_name, instance_namespace);
+
+        let device_usage_state = get_instance_device_usage_state(
+            node_name,
+            instance_name,
+            instance_namespace,
+            &capacity,
+            Arc::new(mock),
+        )
+        .await;
+        assert!(device_usage_state
+            .into_iter()
+            .all(|(_, v)| { v == DeviceUsageStatus::Unknown }));
+    }
+
+    fn create_configuration_device_plugin_service(
+        connectivity_status: InstanceConnectivityStatus,
+        add_to_instance_map: bool,
+    ) -> (DevicePluginService, DevicePluginServiceReceivers) {
+        let (dps, receivers) = create_device_plugin_service(
+            DevicePluginKind::Configuration,
+            connectivity_status,
+            add_to_instance_map,
+        );
+
+        (dps, receivers)
+    }
+
+    // Tests 0 virtual device id is returned if instance not found in InstanceConfig
+    #[tokio::test]
+    async fn test_get_available_virtual_devices_no_instance_in_instance_map() {
+        let _ = env_logger::builder().is_test(true).try_init();
+        let this_node = "node-a";
+        let instance_namespace = "test-namespace";
+        let capacity = 5;
+        let instance_config = InstanceConfigBuilder::new().build();
+        let mock = MockKubeInterface::new();
+
+        let result = get_available_virtual_devices(
+            &instance_config,
+            this_node,
+            instance_namespace,
+            &capacity,
+            Arc::new(mock),
+        )
+        .await;
+        assert!(result.is_empty());
+    }
+
+    // Tests 0 virtual device id is returned if instance not found from kube find_instance
```

Comment:
there are a few tests to validate get_available_virtual_devices that return 0.  it seems like there should be a test that returns an expected non-zero number of results.

Reply:
nevermind ... i see it: test_get_available_virtual_devices_all_free